### PR TITLE
Opencl support

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

**Comment by sergisiso** [comment]
I deleted the add_fortcl and opencl_tmp_fix branches as all the modification are already in this branch.

**Comment by sergisiso** [comment]
@arporter I guess this is next in line to be merged now. All modifications are yours (I just kept it up to date with master and the haloex branch to be used with PSycloneBench but all this is on master now so they don't appear in the PR diff). Is this PR ready in your opinion? Does it make sense that I review it?

**Comment by arporter** [comment]
I think it is ready and yes, please do review it. TBH I've lost track of where we are to a certain extent so I think we just need to push forward to get everything on master and then get everything working with PSycloneBench as a second step.

File: finite_difference/Makefile [hunk 1]
```diff
@@ -61,10 +64,20 @@ dm_fd_lib:
 cl_fd_lib:
 	${MAKE} -C src API_LIB="${API_LIB}" MPI=no OCL=yes
 
+# Target requiring OpenCL installation
+fd_lib_cl: fortcl
```
Comment (sergisiso) [inline finite_difference/Makefile]: What is the difference between this target and the `cl_fd_lib` above? And would it be a lot of work to have a target with MPI=yes and OCL=yes?

File: finite_difference/Makefile [hunk 2]
```diff
@@ -61,10 +64,20 @@ dm_fd_lib:
...
+	${MAKE} -C src API_LIB="${API_LIB}"
+
 install: fd_lib lib include
 	cp src/${API_LIB} ./lib/.
 	cp src/*.[Mm][Oo][Dd] ./include/.
 
+fortcl_stub:
```
Comment (sergisiso) [inline finite_difference/Makefile]: This Makefiles does not use this target

File: finite_difference/Makefile [hunk 3]
```diff
@@ -61,10 +64,20 @@ dm_fd_lib:
...
+	${MAKE} -C ${FORTCL_DIR} stub
+
+fortcl:
```
Comment (sergisiso) [inline finite_difference/Makefile]: This one is used by the `fd_lib_cl`, but I think we could do removing: fortcl, fortcl_stub fd_lib_cl as all this is handled already by the `finite_difference/src/Makefile`. The FORTCL_DIR cleaning I think it is ok.

File: finite_difference/src/Makefile [hunk 4]
```diff
@@ -42,6 +42,9 @@
 #
 # export F90=gfortran
 
+# Location of the FortCL source code (is a git submodule)
```
Comment (sergisiso) [inline finite_difference/src/Makefile]: This is repeated at line 52 and line 100. I think the right one is the declaration at 52.

File: finite_difference/src/Makefile [hunk 5]
```diff
@@ -102,13 +105,15 @@ all: ${API_LIB}
 
 # Create the archive. Note that we include the FortCL objects in this
 # archive but take care to exclude clfortran.o (as that only has
-# interface definitions, not implementation).
+# interface definitions, not implementation). We could use filter-out
+# for this but that is specific to Gnu Make.
 ${API_LIB}: fortcl ${MODULES}
-	${AR} ${ARFLAGS} ${API_LIB} *.o
+	${AR} ${ARFLAGS} ${API_LIB} ${FORTCL_DIR}/*.o ${MODULES}
+	#${AR} ${ARFLAGS} ${API_LIB} *.o
```
Comment (sergisiso) [inline finite_difference/src/Makefile]: This line can be removed

File: finite_difference/src/grid_mod.f90 [hunk 6]
```diff
@@ -244,6 +255,26 @@ function grid_constructor(grid_name, &
...
+    self%dy_f_device = 0
+    self%area_t_device = 0
+    self%area_u_device = 0
+    self%area_v_device = 0
+    self%gphiu_device = 0
+    self%gphiv_device = 0
+    self%gphif_device = 0
+    self%xt_device = 0
+    self%yt_device = 0
+    self%tmask_device = 0
```
Comment (sergisiso) [inline finite_difference/src/grid_mod.f90:276]: Maybe the C binding `c_null_ptr` is more appropriate? and shouldn't the `grid_init` function allocate and initialize these values? (I am not sure when are they allocated at the moment)

**Review by sergisiso — COMMENTED** [review]
@arporter I made a first review of this PR. It is mostly removing some duplications in the build system and some modifications/questions about how device_* are initialised and written.

File: finite_difference/src/field_mod.f90 [hunk 7]
```diff
@@ -291,6 +301,36 @@ end subroutine r2d_free_field
...
+       ! call does nothing.
+       call read_buffer(self%device_ptr, self%data, &
+                        int(self%grid%nx*self%grid%ny, kind=8))
+    end if
+    dptr => self%data
+  end function get_data
+
+  !===================================================
+
+  function set_data(self, array) result(flag)
```
Comment (sergisiso) [inline finite_difference/src/field_mod.f90:323]: Shouldn't this function write to device_ptr if using with the accelerator enabled?

File: finite_difference/Makefile (hunk 1 above)
Comment (arporter) [reply finite_difference/Makefile]: That's a good question :-) I suspect it's old code and I'll remove it. And for your second question, I think that's a separate PR :-)

File: finite_difference/Makefile (hunk 2 above)
Comment (arporter) [reply finite_difference/Makefile]: Removed.

File: finite_difference/Makefile (hunk 3 above)
Comment (arporter) [reply finite_difference/Makefile]: Yes, removed.

File: finite_difference/src/Makefile (hunk 4 above)
Comment (arporter) [reply finite_difference/src/Makefile]: Oops. I've elected to keep the one at 100 as that's consistent with Joerg's changes to allow this library to be built from any location.

File: finite_difference/src/Makefile (hunk 5 above)
Comment (arporter) [reply finite_difference/src/Makefile]: So it can - done.

File: finite_difference/src/grid_mod.f90 (hunk 6 above)
Comment (arporter) [reply finite_difference/src/grid_mod.f90:276]: Oh yes, that's a good idea. However, gfortran then refused to compile the code from which I concluded that these device pointers should be declared as `type(c_ptr)`. If I change that then all of the PSyclone-generated code will probably have to be changed too. Therefore, I suggest we leave it as is for the moment.

Comment (arporter) [reply finite_difference/src/grid_mod.f90:276]: To answer your second question, the initialisation of these pointers happens when required in the code generated by PSyclone - we only do the OpenCL device setup the first time an Invoke is called so we don't have a device until then.

File: finite_difference/src/field_mod.f90 (hunk 7 above)
Comment (arporter) [reply finite_difference/src/field_mod.f90:323]: Yes, good point. However, FortCL has no `write_buffer()` method so I've created an Issue and a TODO for now.

**Comment by arporter** [comment]
I think I've addressed all comments. Ready for 2nd review from @sergisiso

**Comment by sergisiso** [comment]
@arporter I already reviewed this (back in July) and I am happy with how the comments were addressed. I will fix the submodule conflict and since it's been a long time since I reviewed it I will make sure it works with PSyclone and PSycloneBench masters before merging.

**Comment by sergisiso** [comment]
@arporter I checked and it works with PSycloneBench and there is an update needed for PSyclone compilation test which I will make as part of PSyclone/525 PR.

GitHub doesn't let me merge it because some of the changes are made by me. Can you merge this PR into master?

**Review by arporter — APPROVED** [review]
@sergisiso is happy so I'm happy!

**Pull request merged**